src/merkle/tiered_smt/mod.rs

```diff
@@ -166,6 +166,32 @@ impl TieredSmt {
         TieredSmtProof::new(path, entries).expect("Bug detected, TSMT produced invalid proof")
     }
 
+    /// Returns 
+    /// 1. a partial Merkle tree pre-populated with paths leading up to nodes to be inserted,
+    /// one per key in `keys_to_insert`.
+    /// 2. a map `leaf_hash -> (key, value)` which stores the key/value for keys that already contain a value
+    pub fn get_preinsert_partial_merkle_tree<'a>(
+        &'a self,
+        keys_to_insert: impl IntoIterator<Item = &'a RpoDigest>,
+    ) -> Result<(PartialMerkleTree, BTreeMap<RpoDigest, (RpoDigest, Word)>), MerkleError> {
+        let mut pmt = PartialMerkleTree::default();
+        let mut preimage_map = BTreeMap::new();
+
+        for key in keys_to_insert {
+            let (leaf_index, _) = self.nodes.get_leaf_index(key);
```

Comment:
I think we probably do want to use the second return value here. I believe it is a flag which is set to true if the returned index points to a leaf node rather than an internal or an empty node. So, we can use this flag to figure out if we need to add an entry to the `preimage_map`.